# Question 1: Understanding the Classification Problem

## Discussion

The aim of this task is to predict whether a small molecule is toxic or non-toxic to honey bees. The dataset is called ApisTox and contains information about pesticides and their toxicity to honey bees.

The input variables include the source of the chemical data the publication year the strongest toxicity type and information about whether the chemical is used as a herbicide fungicide insecticide or another agrochemical.

The output variable is 'label' . A value of 0 means non toxic and a value of 1 means toxic. Therefore this is a binary classification problem.

This task is useful because toxicity prediction can help with environmental and agricultural research. A wrong prediction can lead to a toxic chemical being treated as safe or a safe chemical being treated as toxic.

Because the two classes are no equal the model should not be judged only by accuracy. Other measures such as precision recall F1score and balanced accuracy should also be used.

In [6]:
import pandas as pd

data = pd.read_csv("dataset_final.csv")

print("Dataset shape:",data.shape)

print("\nTarget distribution:")
print(data["label"].value_counts())

print("\nTarget percentage:")
print(data["label"].value_counts(normalize=True)*100)

print("\nDataset columns:")
print(data.columns.tolist())

Dataset shape: (1035, 13)

Target distribution:
label
0    739
1    296
Name: count, dtype: int64

Target percentage:
label
0    71.400966
1    28.599034
Name: proportion, dtype: float64

Dataset columns:
['name', 'CID', 'CAS', 'SMILES', 'source', 'year', 'toxicity_type', 'herbicide', 'fungicide', 'insecticide', 'other_agrochemical', 'label', 'ppdb_level']


# Question 2: Data Exploration and Preparation

## Discussion

The ApisTox dataset was explored before building the machine learning models. The dataset size, data types, missing values and class distribution were checked.

The dataset contains both numerical and categorical features. Identifier columns and the `ppdb_level` column will not be used as model features. The `ppdb_level` column is closely related to the target label so using it could give the model information about the target.

The dataset does not have missing values. Therefore no missing value replacement is needed. The data will be split into training and testing sets using a 70:30 ratio.


### Data Quality Assessment

The dataset was checked to understand its size its data types missing values and class distribution. This helps us understand the data before preparing it for the machine learning models.

In [7]:
print("Dataset dimensions:")
print(data.shape)

print("\nData types:")
print(data.dtypes)

print("\nMissing values:")
print(data.isnull().sum())

print("\nClass distribution:")
print(data["label"].value_counts())

Dataset dimensions:
(1035, 13)

Data types:
name                  object
CID                    int64
CAS                   object
SMILES                object
source                object
year                   int64
toxicity_type         object
herbicide              int64
fungicide              int64
insecticide            int64
other_agrochemical     int64
label                  int64
ppdb_level             int64
dtype: object

Missing values:
name                  0
CID                   0
CAS                   0
SMILES                0
source                0
year                  0
toxicity_type         0
herbicide             0
fungicide             0
insecticide           0
other_agrochemical    0
label                 0
ppdb_level            0
dtype: int64

Class distribution:
label
0    739
1    296
Name: count, dtype: int64


### Result

The dataset contains 1035 rows and 13 columns. It has both categorical and numerical variables. There are no missing values in the dataset.

The target variable `label` has 739 records in class 0 and 296 records in class 1. This shows that the classes are not balanced. Therefore the class distribution should be considered when evaluating the machine learning models.

## Data Pre-processing

The dataset has no missing values so no missing value treatment is needed. The columns `name`, `CID`, `CAS` and `SMILES` are mainly used to identify or describe the chemical compounds. They were not used as model features.

The `ppdb_level` column was also removed because it is strongly related to the target label and may cause target leakage.

The remaining features contain categorical and numerical data. The data was split into training and testing sets using a 70:30 ratio. Onehot encoding was used for the categorical features and standard scaling was used for the numerical features.